File: src/main/java/org/shsts/tinactory/network/Component.java
package org.shsts.tinactory.network;

import net.minecraft.core.BlockPos;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.BlockState;

import java.util.function.BiConsumer;
import java.util.function.Supplier;

public abstract class Component {
    public interface Factory<T extends Component> {
        T create(ComponentType<T> type, CompositeNetwork network);
    }

    protected final ComponentType<?> type;
    protected final CompositeNetwork network;

    public Component(ComponentType<?> type, CompositeNetwork network) {
        this.type = type;
        this.network = network;
    }

    /**
     * Called during network connect when adding a block to the network.
     */
    public void putBlock(BlockPos pos, BlockState state) {}

    /**
     * Called when network connect is finished.
     */
    public void onConnect() {}

    /**
     * Called when network disconnects.
     */
    public void onDisconnect() {}

    public interface Ticker {
        void tick(Level world, Network network);
    }

    public abstract void buildSchedulings(BiConsumer<Supplier<Scheduling>, Ticker> cons);
}

File: src/test/kotlin/org/shsts/tinactory/datagen/content/component/Component.kt
package org.shsts.tinactory.datagen.content.component

import net.minecraft.world.level.ItemLike
import org.shsts.tinactory.content.electric.Voltage
import java.util.function.Supplier

typealias Component = Map<Voltage, Supplier<out ItemLike>>
typealias NamedComponent = Map<String, Supplier<out ItemLike>>
typealias ListComponent = List<Supplier<out ItemLike>>

fun Component.item(v: Voltage) = getValue(v).get().asItem()

fun ListComponent.item(i: Int) = get(i).get()

fun NamedComponent.item(name: String) = getValue(name).get().asItem()
